### PR TITLE
adding a fully containerized build to the dockerfile

### PR DESCRIPTION
**Pull request by pthomison**
- Copy/Pasta'd my Dockefile golang builder
- Defaulting make build to run tidy as well. I find this useful just to make sure everything is gtg with the mod file
- Adding testing via a container w/ the current directory mounted into it. Primarily doing this so we can use [envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#Environment) to pull down control plane binaries to test API responses for the `kube` package

Will def probably lose some efficiency against using native go, but I have found this handy just to keep all the environment recreation easy. 

If we want, we can keep the default method to be native go + binary COPY && then have the fully containerized build its own option in the Makefile. LMK your thoughts!

### CODE REVIEW

**Comment by pthomison** [comment]
Also dropped the `goldilocks` labels on the dockerfile, not sure if those were a copy/pasta. Can add them back if we want

**Comment by ejether** [comment]
Word! The `golidlocks` tags were total copy-pasta.

File: testing.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+RUN go install sigs.k8s.io/controller-runtime/tools/setup-envtest@latest
+RUN setup-envtest use
+
+WORKDIR /workspace
+# Copy the Go Modules manifests
+COPY go.mod go.mod
+COPY go.sum go.sum
+# cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+RUN go mod download
```
Comment (ejether) [inline testing.Dockerfile:14]:
```suggestion
RUN go mod download

```

File: Makefile [hunk 2]
```diff
@@ -31,7 +31,12 @@ clean:
...
-build-docker: build-linux
+build-docker:
 	docker build -t knamespacer:dev .
+build-docker-test:
+	docker build -t knamespacertest:dev -f testing.Dockerfile .
+docker-test: build-docker-test
+	docker run -it --rm -v "$(shell pwd):$(shell pwd)" -w "$(shell pwd)" knamespacertest:dev \
+	/bin/bash -c "make test"
 # e2e-test:
 # 	venom run e2e/tests/* --output-dir e2e/results --log info --strictf
```
Comment (ejether) [inline Makefile:42]:
```suggestion
# 	venom run e2e/tests/* --output-dir e2e/results --log info --strictf

```

File: Dockerfile [hunk 3]
```diff
@@ -1,14 +1,27 @@
...
+LABEL org.opencontainers.image.licenses="Apache License 2.0"
+WORKDIR /
 # 'nobody' user in alpine
-USER 65534
-COPY knamespacer /
-CMD ["/knamespacer"]
\ No newline at end of file
+USER 65534:65534
+COPY --from=builder /knamespacer/knamespacer .
+ENTRYPOINT ["/knamespacer"]
```
Comment (ejether) [inline Dockerfile:27]:
```suggestion
ENTRYPOINT ["/knamespacer"]

```

File: .gitignore [hunk 4]
```diff
@@ -24,4 +24,8 @@ go.work
 /knamespacer
 
 # Packaged Charts
-knamespacer*.tgz
\ No newline at end of file
+knamespacer*.tgz
+
+# Testing Artifacts
+cover-report.html
+coverage.txt
```
Comment (ejether) [inline .gitignore:31]:
```suggestion
coverage.txt

```

**Review by ejether — CHANGES_REQUESTED** [review]
The multi part build was on my radar, thanks for taking that. 
You didn't add any whitespace errors, but since they show up in this PR, I though, might as well fix them here.

Is the uncomment of the test in the `Makefile` used here for anything. 
I see there are more changes in the next PR.

**Review by ejether — APPROVED** [review]

**Pull request merged**